codegen/src/main.zig

```diff
@@ -700,40 +1282,45 @@ fn reuseCommonType(shape: smithy.ShapeInfo, writer: anytype, state: GenerationSt
     }
     return rc;
 }
-fn shapeInfoForId(id: []const u8, state: GenerationState) !smithy.ShapeInfo {
-    return state.file_state.shapes.get(id) orelse {
+
+fn shapeInfoForId(id: []const u8, shapes: std.StringHashMap(smithy.ShapeInfo)) !smithy.ShapeInfo {
+    return shapes.get(id) orelse {
         std.debug.print("Shape ID not found. This is most likely a bug. Shape ID: {s}\n", .{id});
         return error.InvalidType;
     };
 }
 
+const GenerateTypeOptions = struct {
+    end_structure: bool,
+    key_case: case.Case,
+
+    pub fn endStructure(self: @This(), value: bool) GenerateTypeOptions {
+        return .{
+            .end_structure = value,
+            .key_case = self.key_case,
+        };
+    }
+
+    pub fn keyCase(self: @This(), value: case.Case) GenerateTypeOptions {
+        return .{
+            .end_structure = self.end_structure,
+            .key_case = value,
+        };
+    }
+};
+
 /// return type is anyerror!void as this is a recursive function, so the compiler cannot properly infer error types
-fn generateTypeFor(shape_id: []const u8, writer: anytype, state: GenerationState, end_structure: bool) anyerror!bool {
+fn generateTypeFor(shape_id: []const u8, writer: anytype, state: GenerationState, comptime options: GenerateTypeOptions) anyerror!bool {
+    const end_structure = options.end_structure;
+
     var rc = false;
 
     // We assume it must exist
-    const shape_info = try shapeInfoForId(shape_id, state);
+    const shape_info = try shapeInfoForId(shape_id, state.file_state.shapes);
     const shape = shape_info.shape;
 
     // Check for ourselves up the stack
-    var self_occurences: u8 = 0;
-    for (state.type_stack.items) |i| {
-        // NOTE: shapes.get isn't providing a consistent pointer - is it allocating each time?
-        // we will therefore need to compare ids
-        if (std.mem.eql(u8, i.*.id, shape_info.id))
-            self_occurences = self_occurences + 1;
-    }
-    // Debugging
-    // if (std.mem.eql(u8, shape_info.name, "Expression")) {
-    //     std.log.info("  Type stack len: {d}, occurences: {d}\n", .{ type_stack.items.len, self_occurences });
-    //     if (type_stack.items.len > 15) {
-    //         std.log.info("  Type stack:\n", .{});
-    //         for (type_stack.items) |i|
-    //             std.log.info("  {s}: {*}", .{ i.*.id, i });
-    //         return error.BugDetected;
-    //     }
-    // }
-    // End Debugging
+    const self_occurences: u8 = state.getTypeRecurrenceCount(shape_id);
```

Comment:
Not the point of this PR, but as an aside, I'd love to support the few services that do have these type cycles. It requires some structural changes to how the generation is working